File: Assignment 1/ques1.py
def power(x,y):
    result = 1
    if y==0:
        return 1
    while( y != 0 ):
        result = result * x
        y = y-1
    return result

def nthroot(low,high,n,k):
    if low <= high:
        mid = (low + high) // 2
        if power(mid,k) <= n and power(mid+1,k) > n:
            return mid
        elif power(mid,k) < n:
            return nthroot(mid+1,high,n,k)
        else:
            return nthroot(low, mid-1, n ,k)
    return low


if __name__ == "__main__":
    n = int(input())
    k = int(input())

    result = int(nthroot(0,n,n,k))
    print(result)
    

File: Assignment 1/ques1.cpp
#include <iostream>

using namespace std;

bool bits(unsigned long long int x, unsigned long long int y) {
	unsigned long long int x_bits = 0;
	unsigned long long int n = x;
	while(x){
		x_bits++;
		n = n>>1;
	}
	cout<<x_bits;
	if(x_bits * y > 64){
		return false;
	}
	return true;
}

unsigned long long int power(unsigned long long int x, unsigned long long int y)
{
	unsigned long long int result=1;
	if (y == 0)
		return 1;
	for(; y != 0 ; --y){
		result *= x;
	}
	return result;
}

unsigned long long int nthRootSearch(unsigned long long int low, unsigned long long int high,
						 unsigned long long int N, unsigned long long int K)
{

	if (low <= high)
	{

		unsigned long long int mid = (low + high) / 2;


		if ((power(mid, K) <= N) && bits(mid,K) && (power(mid + 1, K) > N) && bits(mid+1, K))
		{
			return mid;
		}

		else if (bits(mid,K) && power(mid, K) < N)
		{
			return nthRootSearch(mid + 1,
								 high, N, K);
		}
		else
		{
			return nthRootSearch(low,
								 mid - 1,
								 N, K);
		}
	}
	return low;
}

int main()
{

	unsigned long long int N = 10000000000000000;
	unsigned long long int k;

	cin>>k;

	cout << (nthRootSearch(0, N, N, k))
		 << endl;
	return 0;
}
